darity.go

```diff
@@ -7,28 +7,142 @@ package darity
-	ret, _, errnop := syscall.Syscall(
-		syscall.SYS_IOCTL,
-		uintptr(fd),
-		uintptr(request),
-		argp)
-	if errnop != 0 {
-		return 0, os.NewSyscallError("ioctl", fmt.Errorf("%d", int(errnop)))
+const (
+	// Version is the expected KVM API version, returned by APIVersion.  This
+	// version number is taken directly from KVM API documentation, found here:
+	// https://www.kernel.org/doc/Documentation/virtual/kvm/api.txt.
+	//
+	// If KVM does not return this version number when APIVersion is called,
+	// no further interactions should be performed.
+	Version = 12
+
+	// devKVM is the name of the KVM virtual device.
+	devKVM = "/dev/kvm"
+)
+
+var (
+	// ErrIncorrectVersion is returned when an incorrect KVM version is found
+	// when Enabled is called.  From KVM documentation:
+	//   Applications should refuse to run if KVM_GET_API_VERSION returns a
+	//   value other than 12.
+	ErrIncorrectVersion = errors.New("incorrect KVM version")
+)
+
+// Enabled checks for the existence of the KVM virtual device on this system,
+// and also verifies that the KVM version is equal to Version.  If
+// ErrIncorrectVersion is returned, no further KVM interactions should be
+// performed.
+//
+// If the KVM virtual device exists and reports the correct version, boolean
+// true and nil error will be returned.
+//
+// If the KVM device exists, but reports an incorrect version, boolean false
+// and ErrIncorrectVersion will be returned.
+//
+// If the KVM device does not exist, boolean false and nil error will be
+// returned.
+//
+// In addition, other errors may be returned due to failed syscalls.
+func Enabled() (bool, error) {
+	// Verify existence of device
+	exists, err := deviceExists(devKVM)
+	if err != nil {
+		return false, err
 	}
-	return int(ret), nil
+
+	// Device does not exist
+	if !exists {
+		return false, nil
+	}
+
+	// Verify correct KVM API version
+	version, err := APIVersion()
+	if err != nil {
+		return false, err
+	}
+
+	// Incorrect API version
+	if version != Version {
+		return false, ErrIncorrectVersion
+	}
+
+	// Device exists and reports correct API version
+	return true, nil
 }
 
+// APIVersion returns the current KVM API version, as reported by the KVM
+// virtual device.
 func APIVersion() (int, error) {
-	fd, err := syscall.Open("/dev/kvm", syscall.O_RDWR|syscall.O_CLOEXEC, 0)
+	// KVM version number retrieved from closure
+	var version int
+
+	// Invoke closure with KVM device file descriptor to retrieve the
+	// current KVM API version
+	err := withFD(func(fd uintptr) error {
+		v, err := ioctl(fd, C.KVM_GET_API_VERSION, 0)
+		if err != nil {
+			return err
+		}
+
+		version = v
+		return nil
+	})
+
+	return version, err
+}
+
+// withFD opens the KVM virtual device and invokes the input closure, passing in
+// the uintptr file descriptor of the KVM virtual device.  The device is
+// automatically closed on return.
+func withFD(fn func(fd uintptr) error) error {
+	// os.OpenFile sets O_CLOEXEC automatically
```

Comment:
Ah, cool. Can we count on that behaviour, since it's technically undocumented? More importantly, I guess, is there any harm in being explicit?